### DISCUSSION TITLE
Spatial downscaling with Xclim?

### DISCUSSION THREAD

**Discussion by tlsw231** [Ouranosinc/xclim#1584] [Questions]
### Context

This is a very basic question, but does Xclim only support temporal downscaling at this point? I am looking at the documentation and all the examples there appear to showcase downscaling in time. 

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by aulemahal** [comment]
Sadly yes, only the "bias-adjustment" part of "sdba" is implemented.  Currently the module grows according to our needs at Ouranos, but we would be very happy to welcome any contributions! Or at least, if you have some specific examples of algortihms/methods you'd like to see here, we might be able to fit its development somewhere.

**Comment by Zeitsperre** [comment]
I'm not sure if this is what you're looking for, but a number of the developers of `xclim` have also helped develop spatial downscaling/regridding tools in another library: [xesmf](https://xesmf.readthedocs.io/en/latest/).

**Comment by tlsw231** [comment]
Thanks for the prompt replies. So, is one possible route to first bias-adjust using `xclim` and then interpolate the adjustment factors onto the destination grid with higher resolution using `xesmf`?

I am trying to come up with an xarray-based workflow that I can use to downscale CAM-SE output at `ne30` resolution (~ 110 km) to a finer grid.

**Comment by aulemahal** [comment]
That could a be solution! I've never tried, but I have seen articles doing so. And xclim should be able to handle it (see the docs for re-creating a bias-adjustment object from a dataset).

Currently, at Ouranos, we usually regrid the simulation to the reference's grid (or vice-versa) before doing any bias-adjustment. It's not the cleanest, but the method has yielded reasonable results up to now, and it's much simpler.

**Comment by tlsw231** [comment]
Thanks, your solution is so much simpler :) I will check out the docs as you suggest. Do you think it would be worth adding a full example to the `xclim` docs including the interpolation steps back and forth between the simulation and reference grids? This could just be something very similar to what you already seem to be doing.

**Comment by Zeitsperre** [comment]
@tlsw231 We don't have any examples yet of using xesmf with xclim, but if you wanted to add a notebook that employs spatiotemporal downscaling, I think it would be a nice addition to the examples. What do you think @aulemahal ?

**Comment by aulemahal** [comment]
Mentioning in the docs would be a first step. For technical reasons, I'm not sure we should add a fully functional notebook that uses xESMF : we would then need to have it run on the ReadTheDocs CI and xESMF is not lightweight. I think this kind of full workflow should be showcased somewhere else, but I'm not sure where...

My opinion holds as long as we're talking about regridding before and after the bias-adjustment. If one was to develop a method where the trained data (the factors) were regridded between the train and the adjust, then it would merit a notebook that can run on RTD. I'm not planning on working on anything like this soon, but we would be happy to welcome and support any new contributors!

**Comment by tlsw231** [comment]
Thanks for your comments. I am still trying to figure out how to generate the weighting file for unstructured -> regular interpolation using esmpy. Once I have that sorted out, I can gladly share a full workflow.

**Comment by aulemahal** [comment]
If that helps, when using `xESMF` (which I would recommend over calling directly `esmpy`), the only current way to pass non-structured grid is through a "locstream", so a 1D list of points. The support for complex meshes is not yet implemented, mainly because there's no clear way on how to store it in a xarray structure.

**Comment by ArcticSnow** [comment]
Hello @tlsw231 ,
In case this can help you, I gonna point you to the tool I have been part of the development team: https://github.com/ArcticSnow/TopoPyScale

We published a JOSS paper about it: https://doi.org/10.21105/joss.05059
